### PR TITLE
feat: GET /retrieval-success-rate

### PR DESCRIPTION
**Pull request by bajtos**
Implement the first API endpoint for retrieving Retrieval Success Rate.

As part of this work, I created helpers that will make subsequent enpoints
faster to build.

Links:
- https://github.com/filecoin-station/roadmap/issues/57
- https://github.com/filecoin-station/spark-stats/pull/2

### CODE REVIEW

File: lib/handler.js [hunk 1]
```diff
@@ -54,3 +77,105 @@ const notFound = (res) => {
...
+export const today = () => new Date().toISOString().split('T')[0]
+
+/**
+ * @param {string} pathname
+ * @param {string} search
+ * @param {import('node:http').ServerResponse} res
+ * @returns {{from: string | undefined; to: string | undefined}}
+ */
+export const parseAndValidateFilter = (pathname, search, res) => {
+  let { from, to } = qs.parse(search)
```
Comment (juliangruber) [inline lib/handler.js:116]: can be refactored using the `Url` global here as well

File: lib/handler.js [hunk 2]
```diff
@@ -54,3 +77,105 @@ const notFound = (res) => {
...
+}
+
+/**
+ * @param {import('node:http').ServerResponse} res
+ * @param {{from: string, to: string}} filter
+ */
+const setCacheControlForStatsResponse = (res, filter) => {
+  // We cannot simply compare filter.to vs today() because there may be a delay in finalizing
+  // stats for the previous day. Let's allow up to one hour for the finalization.
+  const boundary = new Date(Date.now() - 3600_000).toISOString().split('T')[0]
```
Comment (juliangruber) [inline lib/handler.js:172]: what about moving `d => d.toISOString().split('T')[0]` into its own function? it's also used by `today()`

**Review by juliangruber — CHANGES_REQUESTED** [review]

File: lib/handler.js [hunk 3]
```diff
@@ -28,8 +31,28 @@ export const createHandler = ({
  * @param {import('pg').Pool} pgPool
  */
 const handler = async (req, res, pgPool) => {
-  // TBD
-  notFound(res)
+  const [pathname, search] = parseRequestUrl(req.url)
```
[the remark this answers is not in the document]
Comment (bajtos) [reply lib/handler.js:34]: I tried that, but it does not work :( `req.url` is not a valid URL.

```
> new URL('/foo')
Uncaught TypeError: Invalid URL
    at new URL (node:internal/url:775:36) {
  code: 'ERR_INVALID_URL',
  input: '/foo'
}
```

File: lib/handler.js [hunk 4]
```diff
@@ -54,3 +77,105 @@ const notFound = (res) => {
...
+/**
+ * @param {string} pathname
+ * @param {string} search
+ * @param {import('node:http').ServerResponse} res
+ * @param {string} querystring
+ * @param {import('pg').Pool} pgPool
+ */
+const getRetrievalSuccessRate = async (pathname, search, res, pgPool) => {
+  const filter = parseAndValidateFilter(pathname, search, res)
+  if (res.headersSent) return
```
[the remark this answers is not in the document]
Comment (bajtos) [reply lib/handler.js:90]: Yeah, I agree this is not clear. I'll think a bit more about a better design.

File: lib/handler.js (hunk 1 above)
Comment (bajtos) [reply lib/handler.js:116]: Good point. I see that Url provides `searchParams` with key/value entries.

```
> new URL('http://example.com/foo?key=value')
URL {
  href: 'http://example.com/foo?key=value',
  origin: 'http://example.com',
  protocol: 'http:',
  username: '',
  password: '',
  host: 'example.com',
  hostname: 'example.com',
  port: '',
  pathname: '/foo',
  search: '?key=value',
  searchParams: URLSearchParams { 'key' => 'value' },
  hash: ''
}
```

File: lib/handler.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (bajtos) [reply lib/handler.js]: Fixed in bc6ec03

File: lib/handler.js (hunk 1 above)
Comment (bajtos) [reply lib/handler.js]: See bc6ec03

File: lib/handler.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (bajtos) [reply lib/handler.js]: In 8a20214, I refactored all code not specific to SQL queries into a helper function that wraps all stats-specific handlers now.

**Review by juliangruber — APPROVED** [review]

**Review by juliangruber — CHANGES_REQUESTED** [review]
tests are failing

**Comment by bajtos** [comment]
> tests are failing

Fixed in [fedf821](https://github.com/filecoin-station/spark-stats/pull/4/commits/fedf82144f71a884749435d56c4399be662e0c62) - we need to run the DB schema migration script from spark-evaluate to create `spark_stats` tables.

**Pull request merged**